### PR TITLE
Refactor Logging and Exceptions

### PR DESCRIPTION
**Pull request by iakov-aws**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cid/common.py [hunk 1]
```diff
@@ -162,11 +160,13 @@ def wrap(self, *args, **kwargs):
     def __loadPlugins(self) -> dict:
         if sys.version_info < (3, 8):
             from importlib_metadata import entry_points
```
Comment (darken99) [inline cid/common.py:162]: Seems like this should be moved to the top of the file

File: cid/common.py [hunk 2]
```diff
@@ -567,7 +569,7 @@ def _share(self, dashboard_id, **kwargs):
 
         if dashboard is None:
             print('not deployed.')
-            exit(1)
+            return
```
Comment (darken99) [inline cid/common.py:572]: Why?

File: cid/common.py [hunk 3]
```diff
@@ -554,11 +556,11 @@ def _share(self, dashboard_id, **kwargs):
 
         if not dashboard_id:
             if not self.qs.dashboards:
-                print('\nNo deployed dashboards found')
-                exit()
+                print('No deployed dashboards found')
```
Comment (darken99) [inline cid/common.py:559]: Why?

File: cid/common.py [hunk 4]
```diff
@@ -554,11 +556,11 @@ def _share(self, dashboard_id, **kwargs):
...
+                return
             dashboard_id = self.qs.select_dashboard(force=True)
             if not dashboard_id:
-                exit()
+                return
```
Comment (darken99) [inline cid/common.py:563]: Why?

File: cid/common.py [hunk 5]
```diff
@@ -670,8 +671,11 @@ def _share(self, dashboard_id, **kwargs):
                 })
 
             logger.info(f'Sharing dashboard {dashboard.name} ({dashboard.id})')
-            self.qs.update_dashboard_permissions(DashboardId=dashboard.id, **dashboard_params)
-            logger.info(f'Shared dashboard {dashboard.name} ({dashboard.id})')
+            try:
```
Comment (darken99) [inline cid/common.py:674]: This should be handled inside the `update*` function

File: cid/common.py [hunk 6]
```diff
@@ -718,12 +722,12 @@ def update(self, dashboard_id, recursive=False, force=False, **kwargs):
         if not dashboard_id:
             if not self.qs.dashboards:
                 print('\nNo deployed dashboards found')
-                exit()
+                return
```
Comment (darken99) [inline cid/common.py:725]: Why?

File: cid/common.py [hunk 7]
```diff
@@ -718,12 +722,12 @@ def update(self, dashboard_id, recursive=False, force=False, **kwargs):
...
             dashboard_id = self.qs.select_dashboard(force)
             if not dashboard_id:
                 if not force:
                     print('\nNo updates available or dashboard(s) is/are broken, use --force to allow selection\n')
-                exit()
+                return
```
Comment (darken99) [inline cid/common.py:730]: Why?

File: cid/common.py [hunk 8]
```diff
@@ -744,14 +748,14 @@ def update_dashboard(self, dashboard_id, recursive=False, **kwargs):
                 message=f'Dashboard template changed, update it anyway?',
                 choices=['yes', 'no'],
                 default='yes') != 'yes':
-                exit()
+                return
```
Comment (darken99) [inline cid/common.py:751]: Why?

File: cid/common.py [hunk 9]
```diff
@@ -744,14 +748,14 @@ def update_dashboard(self, dashboard_id, recursive=False, **kwargs):
...
         elif dashboard.latest:
             if get_parameter(
                 param_name=f'confirm-update',
                 message=f'No updates available, should I update it anyway?',
                 choices=['yes', 'no'],
                 default='yes') != 'yes':
-                exit()
+                return
```
Comment (darken99) [inline cid/common.py:758]: Why?

File: cid/common.py [hunk 10]
```diff
@@ -943,79 +946,93 @@ def create_or_update_dataset(self, dataset_definition: dict, dataset_id: str=Non
         cur_required = dataset_definition.get('dependsOn', dict()).get('cur')
         athena_datasource = None
 
-        if not len(self.qs.athena_datasources):
+        if get_parameters().get('quicksight-datasource-id'):
```
Comment (darken99) [inline cid/common.py:949]: Seems like a duplicate from datasources PR

File: cid/utils.py [hunk 11]
```diff
@@ -29,6 +32,18 @@ def intersection(a: Iterable, b: Iterable) -> Iterable:
 def difference(a: Iterable, b: Iterable) -> Iterable:
     return sorted(list(set(a).difference(b)))
 
+def terminate(message: str, code: int=1):
```
Comment (darken99) [inline cid/utils.py:35]: This is not used anywhere

**Review by darken99 — CHANGES_REQUESTED** [review]

File: cid/helpers/athena.py [hunk 12]
```diff
@@ -253,11 +256,9 @@ def execute_query(self, sql_query, sleep_duration=1, database: str=None, catalog
         elif not fail:
             return False
         else:
-            failure_reason = response['QueryExecution']['Status']['StateChangeReason']
-            logger.error('Athena query failed: {}'.format(failure_reason))
-            logger.error(f'Failure reason: {failure_reason}')
-            logger.info('Full query: {}'.format(sql_query))
-            exit(1)
+            logger.error(f'Athena query failed: {failure_reason}')
```
[the remark this answers is not in the document]
Comment (iakov-aws) [reply cid/helpers/athena.py:259]: indeed

File: cfn-templates/cid-cfn.yml [hunk 13]
```diff
@@ -947,6 +948,7 @@ Resources:
               properties = event.get('ResourceProperties', {})
               status, reason = ('FAILED', "Undef error")
               physical_id = properties.get('Dashboard', {}).get('dashboard-id', 'Unknown')
+              region = context.invoked_function_arn.split(":")[3]
```
[the remark this answers is not in the document]
Comment (iakov-aws) [reply cfn-templates/cid-cfn.yml]: yes. THis looks like it is not even needed here

**Review by darken99 — APPROVED** [review]

File: cid/common.py (hunk 5 above)
Comment (iakov-aws) [reply cid/common.py]: Well, i would say that the calling function generally knows the context - if AccessDeniedException can be safely ignored or not.

File: cid/common.py (hunk 1 above)
Comment (iakov-aws) [reply cid/common.py]: yes. we can do that.

Comment (iakov-aws) [reply cid/common.py]: Like that https://github.com/aws-samples/aws-cudos-framework-deployment/pull/383/commits/094a82a1b9d2954fd13193a89e91793048e931ac ?

**Review by darken99 — APPROVED** [review]

**Pull request merged**